src/globals.c

```diff
@@ -212,13 +213,37 @@ bool set_format_string(char *userinput)
 	return true;
 }
 
+bool set_state_dir(char *dir)
```

Comment:
My preference is for set_state_dir() to handle the state_dir global analogous to the way set_format_string() handles format_string. This way, the interface that we use for handling globals will be more consistent.

In this case, I'm thinking init_globals() would call set_state_dir(NULL), which means that either the value passed to --statedir is set, or the default value (STATE_DIR). If the argument to set_state_dir() is not NULL, then we can do validation and set the user-defined value.